gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+
+from gbpservice.neutron.nsf.core import cfg as core_cfg
+from gbpservice.neutron.nsf.core import lb as core_lb
+from gbpservice.neutron.nsf.core import threadpool as core_tp
+from gbpservice.neutron.nsf.core import periodic_task as core_periodic_task
+from gbpservice.neutron.nsf.core import queue as core_queue
+
+from neutron.common import rpc as n_rpc
+
+from oslo_log import log as logging
+from oslo_config import cfg
+from oslo_service import periodic_task as oslo_periodic_task
+from oslo_service import loopingcall
+from oslo_service import service as os_service
+
+LOG = logging.getLogger(__name__)
+
+process = os.getpid()
+
+
+def is_class(obj):
+    return 'class' in str(type(obj))
+
+
+def name(obj):
+    if callable(obj):
+        return "{0}.{1}.{2}".format(
+            type(obj.im_self).__module__,
+            type(obj.im_self).__name__,
+            obj.__name__)
+    elif is_class(obj):
+        return "{0}.{1}".format(
+            type(obj).__module__,
+            type(obj).__name__)
+    else:
+        return obj.__name__
+
+
+def identify(obj):
+    try:
+        return "(%s)" % (name(obj))
+    except:
+        return ""
+
+
+class AgentReportState(object):
+
+    def __init__(self, agent_state):
+        self._context = context.get_admin_context_without_session()
+        self._agent_state = agent_state
+        self._report_topic = agent_state['plugin_topic']
+        self._report_interval = agent_state['report_interval']
+        self._state_rpc = agent_rpc.PluginReportStateAPI(
+            self._report_topic)
+
+    def report_state(self):
+        try:
+            LOG.debug(_("Reporting state ", self._agent_state))
+            self._state_rpc.report_state(self._context, self._agent_state)
+            self._agent_state.pop('start_flag', None)
+        except AttributeError:
+            # This means the server does not support report_state
+            LOG.warn(_("Neutron server does not support state report."
+                       " Agent State reporting will be "
+                       "disabled."))
+            self._heartbeat.stop()
+            return
+        except Exception:
+            LOG.exception(_("Stopped reporting agent state!"))
+
+
+class RpcAgent(n_rpc.Service):
+
+    def __init__(
+            self, sc, host=None,
+            topic=None, manager=None, report_state=None):
+        super(RpcAgent, self).__init__(host=host, topic=topic, manager=manager)
+        if report_state:
+            self._report_state = AgentReportState(self.report_state)
+
+    def start(self):
+        LOG.debug(_("RPCAgent listening on %s" % (self.identify)))
+        super(RpcAgent, self).start()
+
+    def report_state(self):
+        if hasattr(self, '_report_state'):
+            LOG.debug(_("Agent (%s) reporting state" % (self.identify())))
+            self._report_state.report_state()
+
+    def identify(self):
+        return "(host=%s,topic=%s)" % (self.host, self.topic)
+
+
+class RpcAgents(object):
+
```

Comment:
Actually, the process does process RPC calls.. i mean the RPCAgents are invoked from main process and registered to it.
Config Agent has --- a) VPN b) FW c) LB d) SM
Similarly orchestrator and configurator also has RPC endpoints.